zip_file/src/com/google/android/vending/expansion/zipfile/ZipResourceFile.java

```diff
@@ -342,7 +342,8 @@ void addPatchFile(String zipFileName) throws IOException
          * archive. After that, we can release our EOCD hunt buffer.
          */
 
-        int numEntries = bbuf.getShort(eocdIdx + kEOCDNumEntries);
+        // "numEntries" is a 16-bit _unsigned_ value so make it a "char"
+        char numEntries = (char) bbuf.getShort(eocdIdx + kEOCDNumEntries);
```

Comment:
Just to be clear here, the issue is that the short value (`numEntries`) is already overflowed when it's unpacked from `bbuf`, correct? And you're trying to un-overflow it by placing it into an (unsigned) char?

While this is technically correct, I'm not sure I like the idea of (ab)using a char type this way, since it's not character data. Instead, just use a mask to convert it:

```
public static int getUnsignedShort(short x) {
    return x & 0x0000ffffL;
}
```

Reply:
I'm not wed to any particular implementation (nor am I involved with this code any more so I no longer have the means to easily test a different one).